File: lib/features/profitability/domain/profitability_engine.dart
import 'cost_breakdown.dart';
import 'profitability_input.dart';
import '../../profile/domain/fixed_cost_allocation.dart';
import 'cost_settings.dart';

class ProfitabilityEngine {
  CostBreakdown evaluate(ProfitabilityInput input) {
    final distance = input.offer.distanceKm;
    final energyCost = distance *
        (input.vehicle.energyConsumptionPer100Km / 100) *
        input.vehicle.energyPricePerUnit;
    final maintenanceCost = distance * input.vehicle.maintenancePerKm;
    final depreciationCost = distance * input.vehicle.depreciationPerKm;
    final socialContributions =
        input.offer.payoutEuro * input.costs.socialContributionRate;
    final incomeTax =
        input.offer.payoutEuro * (input.costs.incomeTaxRate ?? 0);
    final fixedCostAllocation = _fixedCostAllocation(
      input: input,
      costs: input.costs,
    );
    final totalCosts = energyCost +
        maintenanceCost +
        depreciationCost +
        socialContributions +
        incomeTax +
        fixedCostAllocation;
    final netProfit = input.offer.payoutEuro - totalCosts;

    return CostBreakdown(
      energyCost: energyCost,
      maintenanceCost: maintenanceCost,
      depreciationCost: depreciationCost,
      socialContributions: socialContributions,
      incomeTax: incomeTax,
      fixedCostAllocation: fixedCostAllocation,
      totalCosts: totalCosts,
      netProfit: netProfit,
    );
  }

  double _fixedCostAllocation({
    required ProfitabilityInput input,
    required CostSettings costs,
  }) {
    if (costs.monthlyFixedCosts <= 0) {
      return 0;
    }
    switch (costs.fixedCostAllocation) {
      case FixedCostAllocation.perHour:
        final durationMinutes = input.offer.durationMinutes;
        if (durationMinutes == null || durationMinutes <= 0) {
          throw StateError('Missing offer duration for hourly allocation.');
        }
        if (costs.monthlyWorkingHours <= 0) {
          throw StateError('Monthly working hours must be set.');
        }
        return (costs.monthlyFixedCosts / costs.monthlyWorkingHours) *
            (durationMinutes / 60);
      case FixedCostAllocation.perKm:
        if (costs.monthlyDistanceKm <= 0) {
          throw StateError('Monthly distance must be set.');
        }
        return (costs.monthlyFixedCosts / costs.monthlyDistanceKm) *
            input.offer.distanceKm;
      case FixedCostAllocation.perDelivery:
        if (costs.monthlyDeliveries <= 0) {
          throw StateError('Monthly deliveries must be set.');
        }
        return costs.monthlyFixedCosts / costs.monthlyDeliveries;
    }
  }
}

File: functions/src/profitability_engine.ts
import {
  CostBreakdown,
  CostSettings,
  FixedCostAllocation,
  OfferInput,
  VehicleSnapshot,
} from "./profitability_types";

type ProfitabilityInput = {
  offer: OfferInput;
  vehicle: VehicleSnapshot;
  costs: CostSettings;
};

export function evaluateProfitability(input: ProfitabilityInput): CostBreakdown {
  const distance =
    input.offer.routeVerification?.distanceKm ?? input.offer.distanceKm;
  if (!distance || distance <= 0) {
    throw new Error("Missing offer distance.");
  }
  const energyCost =
    distance *
    (input.vehicle.energyConsumptionPer100Km / 100) *
    input.vehicle.energyPricePerUnit;
  const maintenanceCost = distance * input.vehicle.maintenancePerKm;
  const depreciationCost = distance * input.vehicle.depreciationPerKm;
  const socialContributions =
    input.offer.payoutEuro * input.costs.socialContributionRate;
  const incomeTax = input.offer.payoutEuro * (input.costs.incomeTaxRate ?? 0);
  const fixedCostAllocation = calculateFixedCostAllocation(input, distance);
  const totalCosts =
    energyCost +
    maintenanceCost +
    depreciationCost +
    socialContributions +
    incomeTax +
    fixedCostAllocation;
  const netProfit = input.offer.payoutEuro - totalCosts;

  return {
    energyCost,
    maintenanceCost,
    depreciationCost,
    socialContributions,
    incomeTax,
    fixedCostAllocation,
    totalCosts,
    netProfit,
  };
}

function calculateFixedCostAllocation(
  input: ProfitabilityInput,
  distanceKm: number
): number {
  if (input.costs.monthlyFixedCosts <= 0) {
    return 0;
  }
  switch (input.costs.fixedCostAllocation) {
    case "perHour": {
      const durationMinutes =
        input.offer.routeVerification?.durationMinutes ??
        input.offer.durationMinutes;
      if (!durationMinutes || durationMinutes <= 0) {
        throw new Error("Missing offer duration for hourly allocation.");
      }
      if (input.costs.monthlyWorkingHours <= 0) {
        throw new Error("Monthly working hours must be set.");
      }
      return (
        (input.costs.monthlyFixedCosts / input.costs.monthlyWorkingHours) *
        (durationMinutes / 60)
      );
    }
    case "perKm": {
      if (input.costs.monthlyDistanceKm <= 0) {
        throw new Error("Monthly distance must be set.");
      }
      return (
        (input.costs.monthlyFixedCosts / input.costs.monthlyDistanceKm) *
        distanceKm
      );
    }
    case "perDelivery": {
      if (input.costs.monthlyDeliveries <= 0) {
        throw new Error("Monthly deliveries must be set.");
      }
      return input.costs.monthlyFixedCosts / input.costs.monthlyDeliveries;
    }
    default: {
      return assertUnreachable(input.costs.fixedCostAllocation);
    }
  }
}

function assertUnreachable(value: FixedCostAllocation): number {
  throw new Error(`Unknown fixed cost allocation: ${value}`);
}
